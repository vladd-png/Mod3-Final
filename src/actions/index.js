export const setUrls = (urls) => ({
  type: 'SET_URLS',
  urls
});
