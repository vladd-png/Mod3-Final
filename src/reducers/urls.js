export const urls = (state = [], action) => {
  switch (action.type) {
    case 'SET_URLS':
      return [...state, action.urls];
    default:
      return state;
  }
};
