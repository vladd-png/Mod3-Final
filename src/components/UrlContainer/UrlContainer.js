import React, { Component } from 'react';
import './UrlContainer.css';
import { connect } from 'react-redux';
import { setUrls } from '../../actions';
import { getUrls } from '../../apiCalls';
let urlEls = [];

export class UrlContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      urls: this.props.urls
    }
  }

  componentDidMount() {
    getUrls()
    .then(data => data.urls.forEach((url) => this.props.setUrls(url)))
    .catch(err => console.error('Error fetching:', err))
  }



  render() {
     const urlEls = this.props.urls.map(url => {
       return (
         <div className="url">
           <h3>{url.title}</h3>
           <a href={url.short_url} target="blank">{url.short_url}</a>
           <p>{url.long_url}</p>
         </div>
       )
     })
    return (
      <section>
      { urlEls.length ? urlEls : <p>No urls yet! Find some to shorten!</p> }
      </section>
    )
  }
}

export const mapDispatchToProps = dispatch => {
  return {
    setUrls: urls => dispatch(setUrls(urls))
  }
};

export const mapStateToProps = ({ urls }) => ({
  urls
})

export default connect(mapStateToProps, mapDispatchToProps)(UrlContainer);
