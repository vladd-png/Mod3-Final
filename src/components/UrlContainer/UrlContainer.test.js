import React from 'react';
import { UrlContainer, mapStateToProps, mapDispatchToProps } from './UrlContainer.js'
import { setUrls } from '../../actions';
import { shallow } from 'enzyme';
import { urls } from '../../reducers/urls.js';

// actions
// reducers
// mapStateToProps
// mapDispatchToProps

describe('UrlContainer', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<UrlContainer
        urls={[]}
      />);
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('ACTIONS: should have a type SET_URLS in actions', () => {
    const mockUrl = 'http://mockurlinformation-fortesting.com'
    const expected = {
      type: 'SET_URLS',
      urls: mockUrl
    }
    const result = setUrls(mockUrl)
    expect(result).toEqual(expected)
  })

  it('REDUCER: should return the initial state if no action is provided', () => {
    const expected = []
    const result = urls(undefined, {})
    expect(result).toEqual(expected)
  })

  it('REDUCER: should return the correct state if action type is SET_URLS', () => {
    const mockUrl = {
      id: 2,
      long_url: 'http://mockurlinformation-fortesting.com',
      short_url: 'http://imshort.com',
      title: 'testing'
    }
    const mockState = []
    const mockAction = {
      type: 'SET_URLS',
      urls: mockUrl
    }
    const expected = [{
      id: 2,
      long_url: 'http://mockurlinformation-fortesting.com',
      short_url: 'http://imshort.com',
      title: 'testing'
    }]
    const result = urls(mockState, mockAction)
    expect(result).toEqual(expected)
  })

  it('REDUCER: should be able to add onto state with a new URL when action type is SET_URLS', () => {
    const mockUrl = {
      id: 2,
      long_url: 'http://mockurlinformation-fortesting.com',
      short_url: 'http://imshort.com',
      title: 'testing'
    }
    const mockState = [{
      id: 1,
      long_url: 'http://mockurlinformation-fortesting-butlonger.com',
      short_url: 'http://short.url',
      title: 'testing'
    }]
    const mockAction = {
      type: 'SET_URLS',
      urls: mockUrl
    }
    const expected = [{
      id: 1,
      long_url: 'http://mockurlinformation-fortesting-butlonger.com',
      short_url: 'http://short.url',
      title: 'testing'
    },{
      id: 2,
      long_url: 'http://mockurlinformation-fortesting.com',
      short_url: 'http://imshort.com',
      title: 'testing'
    }]
    const result = urls(mockState, mockAction)
    expect(result).toEqual(expected)
  })

  it('MSTP: should return a url array of urls from redux global store', () => {
    const mockState = { urls: [{
      id: 1,
      long_url: 'http://mockurlinformation-fortesting-butlonger.com',
      short_url: 'http://short.url',
      title: 'testing'
    }],
    doubleCheck: 'should not return me'}
    const expected = { urls: [{
      id: 1,
      long_url: 'http://mockurlinformation-fortesting-butlonger.com',
      short_url: 'http://short.url',
      title: 'testing'
    }]}
    const mappedProps = mapStateToProps(mockState)
    expect(mappedProps).toEqual(expected)
  })

  it('MDTP: should call dispatch when setUrls is called', () => {
    const mockDispatch = jest.fn()
    const mockUrl = 'http://mockurlinformation-fortesting-butlonger.com'
    const actionToDispatch = setUrls(mockUrl)
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.setUrls(mockUrl)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })



})
