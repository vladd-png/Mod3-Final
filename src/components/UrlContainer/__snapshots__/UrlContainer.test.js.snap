// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UrlContainer should match snapshot 1`] = `
<section>
  <p>
    No urls yet! Find some to shorten!
  </p>
</section>
`;
